src/controller/ApplicationController.tsx

```diff
@@ -390,28 +381,91 @@ const ControlledComponent: FC<{
-
-  // If a controller class constructor argument is given then traverse up the
-  // tree until the appropriate controller type is found
-  if (controllerClass) {
-    controller = controller.findControllerInstance(controllerClass);
+function useController(): GenericApplicationController;
+function useController<TController extends ApplicationController>(
+  controllerClass: Constructor<TController>
+): TController;
+function useController<
+  TController extends ApplicationController = ApplicationController,
+>(
+  controllerClass?: Constructor<TController>
+): GenericApplicationController | TController {
+  let controller: GenericApplicationController | null | undefined =
+    useContext(ControllerContext);
+
+  if (controller) {
+    if (controllerClass) {
+      // If a controller class constructor argument is given then traverse up the
+      // tree until the appropriate controller type is found
+      const typedController =
+        controller.findControllerInstance(controllerClass);
+      if (typedController) return typedController;
+    } else {
+      return controller;
+    }
   }
 
-  const statefulController: T = controller;
-  return statefulController;
+  throw new Error(
+    `No controller${controllerClass ? ' of type ' + controllerClass.name : ''} found`
+  );
+}
+
+/**
+ * Accepts the type of a controller or controller constructor and returns the type of its `props` if
+ * possible.
+ */
+export type ControllerProps<TController> =
+  TController extends Constructor<ApplicationController<{}, infer TProps>>
+    ? TProps
+    : TController extends ApplicationController<{}, infer TProps>
+      ? TProps
+      : never;
+
+/**
+ * Get the current `props` of the controller component using context.
+ *
+ * In cases where the controller's props change, this is more efficient than accessing
+ * `controller.props` (i.e. fewer renders) and prevents
+ * [tearing](https://github.com/reactwg/react-18/discussions/69).
+ *
+ * Note that any changed prop will cause a rerender, not only props that are accessed in the
+ * current component. As such, it's even more efficient to simply pass down props manually.
+ *
+ * This is a separate hook because of React's Concurrent features—components are pure and can
+ * safely "see" props from concurrent/abandoned renders, but the controller is stateful and
+ * cannot. So `controller.props` is updated in an Effect only after a render has been committed.
+ *
+ * It's untyped by default but can by typed by any of the following:
+ *
+ * @example
+ * useControllerProps<typeof controller>()
+ * useControllerProps<typeof ControllerClass>()
+ * useControllerProps() as ControllerProps<typeof controller>
+ * useControllerProps() as ControllerProps<typeof ControllerClass>
+ */
+function useControllerProps<TController = unknown>() {
+  const props = useContext(ControllerPropsContext);
+  if (props == null) {
+    throw new Error(`No controller found`);
+  }
+  return props as ControllerProps<TController>;
```

Comment:
I'm not sure we should encourage accessing the controller's props externally. I know we do it, but I think it'd be better as you suggest to just pass props to the subcomponents. `controller.props` is more intended to be used within the controller class itself.

The controller's props are all passed on to the wrapped component itself, so I think rather than providing this hook, we should just suggest grabbing them from there if needed.

Reply:
I'm 100% with you. I'm not trying to encourage this--it's just here to address cases where we already do use `controller.props` in nested components without larger refactors.

Because `controller.props` now has to be updated in an effect, after a render, it causes additional renders and temporary tearing to access them in nested components. The hook is just plain context so you can avoid that issue with a quick refactor. Using context to avoid prop-drilling is what people seem to want anyway in the cases I've seen, even if prop drilling is probably better.